include/sparrow/config.hpp

```diff
@@ -0,0 +1,28 @@
+// Copyright 2024 Man Group Operations Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#if defined(__apple_build_version__)
+#define COMPILING_WITH_APPLE_CLANG
+#endif
+
+consteval bool is_apple_compiler()
+{
+#if defined(__apple_build_version__)
```

Comment:
Suggestion: maybe make `COMPILING_WITH_APPLE_CLANG` be `1` or `0` and then just return here that casted to `bool`? That way we check specifically once if we are using apple clang, but then use the result in the define and here.